rero_mef/authorities/api.py

```diff
@@ -87,13 +87,121 @@ class MefRecord(AuthRecord):
+            search.filter('term', viaf=viaf_pid).source().execute().to_dict()
+        )
+        try:
+            mef_pid = result['hits']['hits'][0]['_source']['pid']
+            return cls.get_record_by_pid(mef_pid)
+        except IndexError:
+            return None
+
+    @classmethod
+    def mef_sync(
+        cls,
+        agency=None,
+        agency_pid=None,
+        viaf_record=None,
+        action=None,
+        delete_pid=True,
+        dbcommit=False,
+        reindex=False,
+        **kwargs
+    ):
+        """Create, update or delete Mef record."""
+        if viaf_record:
+            viaf_pid = viaf_record['viaf_pid']
+            mef_record_from_viaf = cls.get_mef_by_viaf_pid(
+                viaf_pid=viaf_pid, with_deleted=False
+            )
+            mef_record_from_agency = cls.get_mef_by_agency_pid(
+                agency_pid=agency_pid, agency='viaf', with_deleted=False
+            )
+            ref_string = cls.build_ref_string(
+                agency=agency, agency_pid=agency_pid
+            )
+            if action == 'update':
+                if mef_record_from_viaf and mef_record_from_agency:
+                    mef_record_from_viaf.reindex()
+                elif mef_record_from_viaf and not mef_record_from_agency:
+                    mef_record_from_viaf[agency] = {'$ref': ref_string}
+                    mef_record_from_viaf.update(
+                        mef_record_from_viaf,
+                        dbcommit=dbcommit,
+                        reindex=reindex,
+                    )
+                elif not mef_record_from_viaf and not mef_record_from_agency:
+                    from flask import current_app
+                    with current_app.app_context():
+                        url = current_app.config.get('JSONSCHEMAS_HOST')
+                        data = {
+                            '$schema': 'http://' + url + '/schemas/'
+                            'authorities/mef-person-v0.0.1.json',
+                            agency: {'$ref': ref_string},
+                            'viaf': viaf_pid,
+                        }
+                        cls.create(
+                            data,
+                            id_=None,
+                            delete_pid=True,
+                            dbcommit=dbcommit,
+                            reindex=reindex,
+                        )
+                else:
+                    pass
+                    # exception to be throw
+            elif action == 'delete':
+                if mef_record_from_viaf:
+                    mef_record_from_viaf.pop(agency, None)
+                    mef_record_from_viaf.update(
+                        mef_record_from_viaf,
+                        dbcommit=dbcommit,
+                        reindex=reindex,
+                    )
+                # toto: delete mef record if last agency
+            else:
+                print('mef sync action not yet implemented:', action)
```

Comment:
Exception, see https://docs.python.org/3/library/constants.html#NotImplemented for more details

Reply:
done